core-common/src/main/java/org/zalando/nakadi/service/subscription/zk/lock/CuratorNakadiLock.java

```diff
@@ -0,0 +1,87 @@
+package org.zalando.nakadi.service.subscription.zk.lock;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.zalando.nakadi.repository.zookeeper.RotatingCuratorFramework;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Not threads safe, current usage pattern does not require it
+ */
+public class CuratorNakadiLock implements NakadiLock {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(CuratorNakadiLock.class);
+    private static final int SECONDS_TO_WAIT_FOR_LOCK = 15;
+
+    private final RotatingCuratorFramework rotatingCuratorFramework;
+    private final String lockPath;
+
+    private CuratorFramework curatorToRelease;
+    private CuratorFramework curatorLastUsed;
+    private InterProcessSemaphoreMutex semaphore;
+
+    public static String getSubscriptionLockPath(final String subscriptionId) {
+        return "/nakadi/locks/subscription_" + subscriptionId;
+    }
+
+    public CuratorNakadiLock(
+            final RotatingCuratorFramework rotatingCuratorFramework,
+            final String subscriptionId) {
+        this.rotatingCuratorFramework = rotatingCuratorFramework;
+        this.lockPath = getSubscriptionLockPath(subscriptionId);
+    }
+
+    @Override
+    public boolean lock() {
+        try {
+            curatorToRelease = rotatingCuratorFramework
+                    .takeCuratorFramework();
+            // the check is to not recreate semaphore every time
+            if (curatorToRelease != curatorLastUsed) {
+                curatorLastUsed = curatorToRelease;
+                semaphore = new InterProcessSemaphoreMutex(
+                        curatorLastUsed, lockPath);
```

Comment:
On unlock curator is released, therefore semaphore will have wrong reference to curator. I suggest not to cache semaphore as a first step

Reply:
you are right, but semaphore is not used outside of the class and access to it is defined by 2 methods. the next call to lock will replace it. so it will not be used with the same instance of curator.

I have just checked the code of InterProcessSemaphoreMutex, and it does execute any heavy init like creating watchers / znodes. the heavy stuff happens during acquire / release. I think it is good to go with a new instance every time.

Reply:
I decided not to do it and keep it like that for now :)